internal/tiltfile/flags/args_def.go

```diff
@@ -2,66 +2,155 @@ package flags
+		}
+
+		for k, v := range flagsFromArgs {
+			if v.IsSet() {
+				config[k] = v
+			}
+		}
+
+		f, err := os.Create(flagsState.ConfigPath)
+		if err != nil {
+			return nil, false, output, errors.Wrapf(err, "error opening %s for writing", flagsState.ConfigPath)
+		}
+		defer func() {
+			err2 := f.Close()
+			if err2 != nil && err == nil {
+				err = errors.Wrapf(err2, "error closing %s", flagsState.ConfigPath)
+			}
+		}()
+		enc := json.NewEncoder(f)
+		enc.SetIndent("", "  ")
+		err = json.NewEncoder(f).Encode(config)
+		if err != nil {
+			return nil, false, output, errors.Wrapf(err, "error serializing config to %s", flagsState.ConfigPath)
+		}
+
+		mergedArgs = true
+	}
+
+	ret := starlark.NewDict(len(ad.args))
+	for k, v := range config {
+		err := ret.SetKey(starlark.String(k), v.starlark())
+		if err != nil {
+			return nil, false, output, err
+		}
+	}
+
+	return ret, mergedArgs, output, nil
+}
+
+func (ad ArgsDef) parseArgs(args []string) (ret map[string]argValue, output string, err error) {
 	fs := flag.NewFlagSet("", flag.ContinueOnError)
 	w := &bytes.Buffer{}
 	fs.SetOutput(w)
 
+	ret = make(map[string]argValue)
 	for name, def := range ad.args {
+		ret[name] = def.newValue()
 		if name == ad.positionalArgName {
 			continue
 		}
-		v := def.flag()
-		fs.Var(v, name, def.usage)
+		fs.Var(ret[name], name, def.usage)
 	}
 
 	err = fs.Parse(args)
 	if err != nil {
-		return starlark.None, w.String(), err
+		return nil, w.String(), err
 	}
 
 	if len(fs.Args()) > 0 {
 		if ad.positionalArgName == "" {
-			return starlark.None, w.String(), errors.New("positional args were specified, but none were expected (no arg defined with args=True)")
+			return nil, w.String(), errors.New("positional args were specified, but none were expected (no arg defined with args=True)")
 		} else {
-			ad.args[ad.positionalArgName].setFromArgs(fs.Args())
+			ret[ad.positionalArgName].setFromArgs(fs.Args())
 		}
 	}
 
-	ret := starlark.NewDict(len(ad.args))
-	for name, def := range ad.args {
-		err := ret.SetKey(starlark.String(name), def.starlark())
-		if err != nil {
-			return starlark.None, w.String(), err
+	return ret, w.String(), nil
+}
+
+func (ad ArgsDef) readFromFile(tiltConfigPath string) (ret map[string]argValue, err error) {
+	ret = make(map[string]argValue)
+	r, err := os.Open(tiltConfigPath)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return ret, nil
 		}
+		return nil, errors.Wrapf(err, "error opening %s", tiltConfigPath)
 	}
+	defer func() {
+		err2 := r.Close()
+		if err2 != nil && err == nil {
+			err = errors.Wrapf(err2, "error closing %s", tiltConfigPath)
+		}
+	}()
 
-	return ret, w.String(), nil
+	m := make(map[string]interface{})
+	err = jsoniter.NewDecoder(r).Decode(&m)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error parsing json from %s", tiltConfigPath)
+	}
+
+	for k, v := range m {
+		def, ok := ad.args[k]
+		if !ok {
+			return nil, fmt.Errorf("%s specified unknown flag name '%s'", tiltConfigPath, k)
```

Comment:
are there any tests that exercise these errors? i'm not totally sure i understand when these get triggered

Reply:
missed those. added. thanks.